core/src/pcap.rs

```diff
@@ -128,45 +128,102 @@ impl Drop for Pcap {
+            Pcap::create(&format_pcap_file(port_name, core.raw(), "rx"))?;
+            unsafe {
+                ffi::rte_eth_add_rx_callback(
+                    port_id.raw(),
+                    rxq.raw(),
+                    Some(append_and_write_rx),
+                    port_name.to_cstring().into_raw() as *mut raw::c_void,
+                );
+            }
+        }
+        RxTxQueue::Tx(txq) => {
+            Pcap::create(&format_pcap_file(port_name, core.raw(), "tx"))?;
+            unsafe {
+                ffi::rte_eth_add_tx_callback(
+                    port_id.raw(),
+                    txq.raw(),
+                    Some(append_and_write_tx),
+                    port_name.to_cstring().into_raw() as *mut raw::c_void,
+                );
+            }
+        }
+    }
+
     Ok(())
 }
 
-/// Append and write slice of packets to an already-created file, logging errors
-/// that may occur.
-///
-/// # Example
-///
-/// ```
-/// pcap::append_and_write(self.port_id, CoreId::current(), "rx", &ptrs);
-/// ```
-pub(crate) fn append_and_write(
-    port: PortId,
-    core: CoreId,
-    tx_or_rx: &str,
-    ptrs: &[*mut ffi::rte_mbuf],
-) {
-    let path = String::from(format!("{:?}-{:?}-{}.pcap", port, core, tx_or_rx).as_str());
-    let _ = Pcap::append(path.as_str())
-        .map_err(|err| {
-            error!(
-                message = "can't append to pcap file",
-                pcap = path.as_str(),
-                ?err
-            )
-        })
-        .map(|pcap| {
-            pcap.write(&ptrs).map_err(|err| {
+/// Callback fn used in `rte_eth_add_rx_callback`, which is called on RX
+/// with a burst of packets that have been received on a given port and queue.
+unsafe extern "C" fn append_and_write_rx(
+    _port_id: u16,
+    _queue_id: u16,
+    pkts: *mut *mut ffi::rte_mbuf,
+    num_pkts: u16,
+    _max_pkts: u16,
+    user_param: *mut raw::c_void,
+) -> u16 {
+    append_and_write(
+        user_param.as_str(),
+        "rx",
+        std::slice::from_raw_parts_mut(pkts, num_pkts as usize),
+    );
+    num_pkts
+}
+
+/// Callback fn used in `rte_eth_add_tx_callback`, which is called on TX
+/// with a burst of packets immediately before the packets are put onto
+/// the hardware queue for transmission.
+unsafe extern "C" fn append_and_write_tx(
+    _port_id: u16,
+    _queue_id: u16,
+    pkts: *mut *mut ffi::rte_mbuf,
+    num_pkts: u16,
+    user_param: *mut raw::c_void,
+) -> u16 {
+    append_and_write(
+        user_param.as_str(),
+        "tx",
+        std::slice::from_raw_parts_mut(pkts, num_pkts as usize),
+    );
+    num_pkts
+}
+
+/// Executed within the rx/tx callback functions for writing out to pcap
+/// file(s).
+fn append_and_write(port: &str, tx_or_rx: &str, ptrs: &[*mut ffi::rte_mbuf]) {
+    let path = format_pcap_file(port, CoreId::current().raw(), tx_or_rx);
+    match Pcap::append(path.as_str()) {
+        Ok(pcap) => {
+            let _ = pcap.write(&ptrs).map_err(|err| {
```

Comment:
hmm, still calling `map_err` when we are not mapping the error. I think this might be a little more expressive of what's going on

```
if let Err(e) = Pcap::append(path.as_str()).and_then(|pcap| pcap.write(&ptrs)) {
    error!(...)
}
```